src/server/game/Entities/Object/Object.cpp

```diff
@@ -2305,13 +2305,90 @@ void WorldObject::MovePosition(Position &pos, float dist, float angle)
     pos.SetOrientation(GetOrientation());
 }
 
+// Only call this function when you already checked for collision with vmap/gameobject.
+// You should check for collision again after this one has been called.
+// excludeCollisionHeight should only be true if you had collision, it wont add it to raycasts for dest position.
+float WorldObject::SelectBestZForDestination(float x, float y, float z, bool excludeCollisionHeight) const
+{
+    if (Unit const* unit = ToUnit())
+        if (unit->CanFly())
+            return z;
+
+    float myX, myY, myZ;
+    GetPosition(myX, myY, myZ);
+
+    float const myCollisionHeight = excludeCollisionHeight ? 0.0f : GetCollisionHeight();
+
+    float const myGridHeight = GetMap()->GetGridMapHeight(myX, myY);
+    float const myVmapFloor = std::max(GetMap()->GetVMapFloor(myX, myY, myZ, 150.0f, GetCollisionHeight()),
+        GetMap()->GetGameObjectFloor(GetPhaseMask(), myX, myY, myZ, 150.0f, GetCollisionHeight()));
+
+    // which of these 3 do I want ?
+    float const destGridHeight = GetMap()->GetGridMapHeight(x, y);
+    float const destCeil = GetMap()->GetCeil(GetPhaseMask(), x, y, z, 150.0f, myCollisionHeight);
+    float const destVmapFloor = std::max(GetMap()->GetVMapFloor(x, y, z, 150.0f, myCollisionHeight),
+        GetMap()->GetGameObjectFloor(GetPhaseMask(), x, y, z, 150.0f, myCollisionHeight));
+
+    bool const hasVmapFloor = myVmapFloor > INVALID_HEIGHT;
+    bool const hasDestGridHeight = destGridHeight > INVALID_HEIGHT;
+    bool const hasDestVmapCeil = destCeil < VMAP_INVALID_CEIL && destCeil != destVmapFloor;
+    bool const hasDestVmapFloor = destVmapFloor > INVALID_HEIGHT;
+    bool const destBetweenVmaps = hasDestVmapCeil && hasDestVmapFloor;
+    bool const underVmap = !hasDestVmapFloor;
+    bool const aboveAllVmap = !hasDestVmapCeil;
+
+    bool const isOnVmap = hasVmapFloor &&
+        ((myZ < myGridHeight && std::fabs(myVmapFloor - myZ) < std::fabs(myGridHeight - myZ)) ||
+        (myZ > myGridHeight && myVmapFloor > myGridHeight));
+
+    bool const hasToFollowGridHeight = hasDestGridHeight && ((underVmap && aboveAllVmap) ||
```

Comment:
help :'(

Reply:
ugh

A brief description on whats being decided here might be helpful

Reply:
oh boi.
I will see if I can get back to paint + provide in game experiences

Reply:
![image](https://user-images.githubusercontent.com/6825243/35487223-75058728-0479-11e8-946c-9e9adf03465f.png)
We want to follow grid height in this example. purple is vmap/gob height data and green is grid

So, we have a destVmapFloor as you can see, which, for this example, is closer to the vmap than to the grid. Obviously we don't want to end there and we should select gridheight based on these conditions:
- We have a destvmap floor, but are not on a vmap currently.
- There is no vmapceil
- The grid height is higher than the vmap floor we found.

It's not finished, and I'm afraid it never will. But I think you can follow the other cases.
It tried to note everything but got quickly lost in everything and didn't really find a good way to keep track of it.

Tbh, I want this gone. The only tools I have is some modified .gps command which returns the value I want, and the debugger and searching for every corner case is such a fucking pain.